gridlab-d/gridlab-d
# Feature/1060

#### What's this Pull Request do?
This pull request adds a MySQL implementation of the group_recorder class as requested in #1060, as well as tests and support components needed to make the implementation viable. 

#### Where should the reviewer start?
The reviewer will need to ensure gridlab-d is built and installed with MySQL support, and that the MySQL Connector C is installed and linked on their machine. 
To verify correct configuration and system compatibility, the "validate.no" file in the "mysql" directory should be removed, and gridlab-d's validation and test suite should be run. 

#### How should this be tested?
The system should be run using the MySQL module to replace the tape module for several group recorders of varying size, interval, and output count, to verify backwards compatibility with existing group recorders.
Additionally, the system should be tested with bespoke group recorders designed to test the various system functionalities and input options as listed on the [MySQL Group Recorder Wiki Page](http://gridlab-d.shoutwiki.com/wiki/Group_recorder_(mysql)).
Any test cases which might push the edge cases of the system should be run, and verified to either work, or fail in a predictable and resolvable way. 

#### Any background context you want to provide?
#### What are the relevant issues?
MySQL has a hard row-size limit can cause the application to fail out during startup if the simulation produces more than 4096 columns or 65,535 bytes, whichever is smaller. This issue is generally addressed by the auto-set column_limit value, however it is possible to manually increase the value to levels where the system will fault.
Due to the overhead of initializing tables, and transferring data to MySQL, this implementation is slightly slower than the tape group recorder under most conditions, however if the simulation is small enough, and the query_buffer_limit is set low enough, the database interaction overhead can dominate the runtime, and significantly hinder use. 

#### Screenshots (if appropriate)
#### Questions:

[validate.txt](https://github.com/gridlab-d/gridlab-d/files/1669508/validate.txt)
